test/utils/controller_manager.go

```diff
@@ -0,0 +1,108 @@
+	"github.com/kong/kubernetes-testing-framework/pkg/clusters/addons/kong"
+	"github.com/kong/kubernetes-testing-framework/pkg/utils/kubernetes/generators"
+)
+
+// -----------------------------------------------------------------------------
+// Testing Utility Functions - Controller Manager
+// -----------------------------------------------------------------------------
+
+// crds provides the filesystem paths to all the base CRDs needed for controller
+// manager functionality.
+var crds = []string{
+	"../../config/crd/bases/configuration.konghq.com_udpingresses.yaml",
+	"../../config/crd/bases/configuration.konghq.com_tcpingresses.yaml",
+	"../../config/crd/bases/configuration.konghq.com_kongplugins.yaml",
+	"../../config/crd/bases/configuration.konghq.com_kongingresses.yaml",
+	"../../config/crd/bases/configuration.konghq.com_kongconsumers.yaml",
+	"../../config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml",
+}
+
+// DeployControllerManagerForCluster deploys all the base CRDs needed for the
+// controller manager to function, and also runs a copy of the controller
+// manager on a provided test cluster.
+//
+// Controller managers started this way will run in the background in a goroutine:
+// The caller must use the provided context.Context to stop the controller manager
+// from running when they're done with it.
+func DeployControllerManagerForCluster(ctx context.Context, cluster clusters.Cluster, additionalFlags ...string) error {
+	// ensure that the provided test cluster has a kongAddon deployed to it
+	var kongAddon *kong.Addon
+	for _, addon := range cluster.ListAddons() {
+		if addon.Name() == kong.AddonName {
+			var ok bool
+			kongAddon, ok = addon.(*kong.Addon)
+			if !ok {
+				return fmt.Errorf("an invalid kong addon was present in test environment")
+			}
+		}
+	}
+	if kongAddon == nil {
+		return fmt.Errorf("no kong addon found loaded in cluster %s", cluster.Name())
+	}
+
+	// determine the proxy admin URL for the Kong Gateway on this cluster:
+	proxyAdminURL, err := kongAddon.ProxyAdminURL(ctx, cluster)
+	if err != nil {
+		return fmt.Errorf("couldn't determine Kong Gateway Admin URL on cluster %s: %w", cluster.Name(), err)
+	}
+
+	// create a tempfile to hold the cluster kubeconfig that will be used for the controller
+	// generate a temporary kubeconfig since we're going to be using the helm CLI
+	kubeconfig, err := generators.TempKubeconfig(cluster)
+	if err != nil {
+		return err
+	}
+
+	// deploy our CRDs to the cluster
+	for _, crd := range crds {
+		cmd := exec.CommandContext(ctx, "kubectl", "--kubeconfig", kubeconfig.Name(), "apply", "-f", crd) //nolint:gosec
+		stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
+		cmd.Stdout = stdout
+		cmd.Stderr = stderr
+		if err := cmd.Run(); err != nil {
+			os.Remove(kubeconfig.Name())
+			return fmt.Errorf("%s: %w", stderr.String(), err)
+		}
+	}
+
+	// render all controller manager flag options
+	controllerManagerFlags := []string{
```

Comment:
Why do we want to split the controller manager flags into two sets that we merge later? I'm not seeing an obvious reason to do that and, and it can introduce additional burden on the reader, and the risk of flag duplication/shadowing.

Reply:
This was done to provide a default set of flags here for things that are either really basic and fundamental or that are derived from other values in the function (e.g. `--kubeconfig` being derived from the file that was created as part of this function.

Reply:
There is no complexity (that's apparent to me) coming from setting all flags in one place (as it was before). This split introduces complexity though - the reader needs to read in two different functions, and also needs to ensure that they don't duplicate the entries.

Not a blocker but my personal feedback is that I liked the old version more.

Reply:
Complexity no, but some callers may not want all the settings that our integration test suite has, the immediate effect of this change is that the integration test suite uses the exact same flags it did before for the controller manager, but the Istio test only uses a reduced set of defaults needed for the core functionality. Does that make sense, or change your mind at all?

Reply:
I think that sooner or later we'll have a bug where either one flag shadows another, or these two sets of flags interact in an unexpected manner, or somebody makes a change to all tests where they only meant one.

If you think that's a reasonable tradeoff, go for it.

Reply:
I do, thanks for the flexibility :+1: